[16.0] [REF] account_invoice_triple_discount Consolidate discount in std field (Follow-up of #1638)

legalsylvain opened the pull request:
Finish https://github.com/OCA/account-invoicing/pull/1638

Basically 
- **Get all the commits from the previous @grindtildeath PR.**

- **remove** this commit https://github.com/OCA/account-invoicing/pull/1638/commits/01df667c1c3fdae39671a5effc0cb54ef6700fd5 that was introducing incompatibility between ``account_invoice_triple_discount`` and ``account_invoice_global_discount`` that can works together in fact. 

- **Add** 
    - e03dba34: a simple commit to make ``account_global_discount`` working if ``account_invoice_triple_discount`` is installed. With that commit, the [remark of @pedrobaeza](https://github.com/OCA/account-invoicing/pull/1638#issuecomment-2342817865) should be answered.
    - 37c8e087: another commit to exclude mutually ``account_invoice_triple_discount`` and ``account_invoice_fixed_discount``. This commit is a fix in fact, as the exclusion [is written in the documentation](https://github.com/OCA/account-invoicing/tree/14.0/account_invoice_fixed_discount#installation), but the CI and the copier answers don't mention that exclusion.


Let's see if all CI is green.

grindtildeath committed:
[REF] account_invoice_triple_discount: Consolidate discount in std field

Until now, the triple discount feature did use the standard Odoo field
as the first discount field, adding only extra fields for the second
and the third discount. This implied we had to override any function
using discount to consider the other discounts properly.

By adding an extra field discount1 to store the first discount, it allows
to redefine the standard discount field to a computed field that will
consolidate the triple discount from the other fields, and avoid the
need to redefine anything relying on the discount field as it will
already consider the triple discounts.

legalsylvain committed:
[IMP] account_invoice_triple_discount: Simplify account.move.line algorithm

[FIX] account_invoice_triple_discount: populate discount1 field to avoid inconsistency if database contains account.move.line with not null discount before the installation of the module

[REF] account_invoice_triple_discount: Add GRAP as coauthor of the module

[FIX] account_invoice_triple_discount: Manage Correct creation of account move lines, where 'discount' field is set

ramiadavid committed:
Add test

grindtildeath committed:
Write discount into discount 1 if not defined

In case this module is installed and sale module is also installed
but not sale_triple_discount, or in case sale_triple_discount is not
updated to its latest version, make sure that setting discount, without
having discount1 defined, will not mess with the triple discounts.

legalsylvain committed:
[REF] account_global_discount: adapt test to make it working if account_invoice_triple_discount is installed

legalsylvain committed:
[FIX] account_invoice_fixed_discount, account_invoice_triple_discount:
- set modules as incompatible in the manifest. (For the time being, there is only a informative message in the README of the module account_invoice_fixed_discount)
- Run separated tests for both modules.

legalsylvain commented:
CI is green ! 
The original PR was reviewed by @Feyensv, @chienandalu, @lmignon, @sergiocorato, @ramiadavid, @leemannd,  @grindtildeath, @pedrobaeza.  could you take a look ?

pedrobaeza reviewed (approved):
I still don't like incompatibilities/exclusion between modules, but in this case, it seems unavoidable (well, not worth to avoid it in fact).

lmignon commented on account_invoice_triple_discount/migrations/16.0.2.0.0/post-migrate.py:
```diff
@@ -0,0 +1,41 @@
+
+
+@openupgrade.logging()
+def compute_discount(env):
+    move_lines_to_compute = env["account.move.line"].search(
+        [
+            "|",
+            "|",
+            ("discount1", "!=", 0),
+            ("discount2", "!=", 0),
+            ("discount3", "!=", 0),
+        ]
+    )
+    for line in move_lines_to_compute:
+        discount = line._get_aggregated_multiple_discounts(
+            [line[x] for x in ["discount1", "discount2", "discount3"]]
+        )
+        rounded_discount = line._fields["discount"].convert_to_column(discount, line)
+        openupgrade.logged_query(
+            env.cr,
+            """
+            UPDATE account_move_line
+            SET discount = %s
+            WHERE id = %s;
+        """,
+            tuple(
+                [
+                    rounded_discount,
+                    line.id,
+                ]
+            ),
+        )
+
```
This migration script is unusable on a large database :sob: The value should be updated in one single SQL query.

legalsylvain committed:
[FIX] account_invoice_triple_discount: remove digits on overloaded field 'discount' as the field is now a technical field that should not have limitation. For exemple in the default case where digits precision of Discount is 2, if discounts are 05%, 09% and 13% the main discount is 24.7885 % (and not 24.79)

legalsylvain committed:
[REF] account_invoice_triple_discount: Improve speed computation of the field 'discount' and reduce complexity

legalsylvain commented:
Hi. 

> I still don't like incompatibilities/exclusion between modules, but in this case, it seems unavoidable (well, not worth to avoid it in fact).

Well, that PR doesn't change that point. (the incompatibility exists for the time being). It is just mandatory to explicit the incompatibility in the CI, because tests are now failing.

@grindtildeath : There was a problem in the previous implementation, regarding the digits precision. I fixed it here https://github.com/OCA/account-invoicing/pull/1840/commits/62d5d3f4b44a5cc603dda1028c333a4a2ca208b3, could you take a look ? 

@lmignon: I replaced the migration script by a unique SQL request.

### Data before the migration (16.0.1.0.3)

<code>
 id | quantity | price_unit | discount | discount2 | discount3 | price_subtotal 
----+----------+------------+----------+-----------+-----------+----------------
 40 |     1.00 |    1000.00 |    10.00 |     20.00 |     30.00 |         504.00
 43 |     1.00 |    1000.00 |     5.00 |      9.00 |     13.00 |         752.12
 44 |     1.00 |    1000.00 |     1.11 |      2.22 |      3.33 |         934.75

</code>

### Data after the migration (16.0.2.0.0)

<code>
 id | quantity | price_unit |      discount      | discount1 | discount2 | discount3 | price_subtotal 
----+----------+------------+--------------------+-----------+-----------+-----------+----------------
 40 |     1.00 |    1000.00 |               49.6 |     10.00 |     20.00 |     30.00 |         504.00
 43 |     1.00 |    1000.00 |            24.7885 |      5.00 |      9.00 |     13.00 |         752.12
 44 |     1.00 |    1000.00 |       6.5252895786 |      1.11 |      2.22 |      3.33 |         934.75

</code>

lmignon commented on account_invoice_triple_discount/migrations/16.0.2.0.0/pre-migrate.py:
```diff
@@ -0,0 +1,47 @@
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE account_move_line
```
@legalsylvain Your script doesn't take into account the discount_type to compute the discount value....

legalsylvain replied:
Hi @lmignon. could you elaborate ? I don't understand what is discount_type.
**In V16 :** 
<code>
$ rgrep "discount_type" ./ --include=*.py
./addons/l10n_it_edi/models/account_edi_format.py: discount_type = discount_element.xpath('.//Tipo')
./addons/l10n_it_edi/models/account_edi_format.py: if discount_type and discount_type[0].text == 'MG':
./addons/l10n_it_edi/models/account_edi_format.py: discount_type = discount_element.xpath('.//Tipo')
./addons/l10n_it_edi/models/account_edi_format.py: if discount_type and discount_type[0].text == 'MG':
./addons/l10n_it_edi/models/account_invoice.py: def discount_type(discount):
./addons/l10n_it_edi/models/account_invoice.py: 'discount_type': discount_type,
</code>

lmignon replied:
Sorry... it's not in this addon... it's into sale_triple_discount https://github.com/OCA/sale-workflow/blob/16.0/sale_triple_discount/models/sale_order_line.py#L26

legalsylvain commented:
@grindtildeath : could you take a look here, as it is your original work ? 
Thanks !

grindtildeath reviewed (approved):
Thanks for taking this over @legalsylvain :hugs: 

I'll be on holidays for the next 10 days so feel free to merge together with sale_triple_discount refactor.

legalsylvain commented:
I have no right to merge this one. @OCA/accounting-maintainers : could you merge this one ? 

Thanks !

pedrobaeza commented:
/ocabot merge major

legalsylvain commented:
@pedrobaeza. I just rebased this PR and the test are all green. could you try to merge again ? 

sorry for disturbing you, I have no right on accounting repo for the time being.

pedrobaeza commented:
/ocabot merge major

legalsylvain commented:
Hi. i don't understand the CI for the time being.

**Step to reproduce :** 
- click on the link provided by the bot (https://github.com/OCA/account-invoicing/commits/16.0-ocabot-merge-pr-1840-by-pedrobaeza-bump-major)

![image](https://github.com/user-attachments/assets/c11ac918-50df-42d0-9763-0a346028719b)
Until ...
![image](https://github.com/user-attachments/assets/32995966-febc-4a4f-9c38-32453049cdc5)

(82 bot cmmits, 3 pages)


**Result:** 
1) The quantity of "pot" commit is very weird : 

- https://github.com/OCA/account-invoicing/commit/b08a2e53397211b5c22d96a6a9eab7fc3d18f387
- https://github.com/OCA/account-invoicing/commit/b6f010ad02de572fc19acaedf062680492ade6c9

It is always the same diff.

2) Execution of merge is very slow. (pedro asked for the merge 6 hours ago...)

3) Edit : in fact, bot adds new commits each few minutes. As a result quantity of commits is growing, and merge is never done.

CC : @OCA/core-maintainers

pedrobaeza commented:
That's because of the rebel modules, that export the POT separately. Thus, one CI adds some texts, and the other removes them, so this can't be merged unless you compatibilize them and put them all together.

legalsylvain commented:
> That's because of the rebel modules, that export the POT separately. Thus, one CI adds some texts, and the other removes them.

That's looks like a CI problem. How can we fix it ? CC : @sbidoul

could we merge this PR manually in the meantime ? 
Thanks !

legalsylvain committed:
[REF] account_invoice_supplierinfo_update_discount: adapt code to discount refactoring

legalsylvain commented:
hi @OCA/accounting-maintainers : @sbidoul took a look on the bug present on that PR, [here](https://github.com/OCA/oca-addons-repo-template/issues/294#issuecomment-2679030152).
I adapted the copier answer file.

Coud you try to merge again this approved PR ? 

Thanks !

sbidoul commented:
Retrying the merge as requested (I have no opinion on the PR itself).

/ocabot merge major

legalsylvain commented:
@grindtildeath ! finally https://github.com/OCA/account-invoicing/pull/1638 is merged ! Thanks a lot for your huge refactoring and simplification.

@sbidoul thank you for helping me solve this rebellious CI.

manuelregidor commented:
@legalsylvain  I think this change is causing trouble in sale-workflow repository.

lk-eska commented:
@legalsylvain do you have plan to apply the same logic for sale  triple discount module?
https://github.com/OCA/sale-workflow/tree/16.0/sale_triple_discount

legalsylvain commented:
Hi @lk-eska . there is pending work here https://github.com/OCA/sale-workflow/pull/2885 by @grindtildeath
